### DISCUSSION TITLE
Run migrations on NativeDatabase ?

### DISCUSSION THREAD

**Discussion by wstrange** [simolus3/drift#2532] [Q&A] [answered]
This feels like a dumb question, but how to I trigger migrations on a NativeDatabase instance?

This is a server app running on Mac OS, *not* a mobile sqlite instance.

The database file does not seem to contain a version table, so how can the app know the 'current' version number to trigger a migration? 

Bumping `  int get schemaVersion => 2;`.  doesn't do anything.

**Comment by wstrange** [comment]
This is what I hacked together. Not sure if this correct:

```dart

Database() : super(_openConnection()) {
    var m = createMigrator();
    // "upgrade" from version 1 to current schema version
    migration.onUpgrade(m,1,schemaVersion);
  }

```

**Comment by frankvollebregt** [comment]
If you haven't read it yet, perhaps the documentation page can help you out: https://drift.simonbinder.eu/docs/advanced-features/migrations/

That is, you should be able to specify the current schema version in your main database class:
```dart
@override
int get schemaVersion => 42;
```
And when it has changed compared to the last version that was run, the `migration` is automatically run when your program launches:
```dart
@override
MigrationStrategy get migration {
  return MigrationStrategy(
    onCreate: (Migrator m) async {
      await m.createAll();
    },
    onUpgrade: (Migrator m, int from, int to) async {
      if (from < 41) {
        // a column was added
        await m.addColumn(myTable, myTable.myColumn);
      }
      if (from < 42) {
        // some other change
      }
    },
  );
}
```
So simply increasing the `schemaVersion` should trigger a migration to run

**Reply by simolus3** [reply]
I agree with this advice. Adding to it, drift migrations are based on the [user-version](https://www.sqlite.org/pragma.html#pragma_user_version) pragma stored in the header of the sqlite3 file. This works on mobile and for servers, so there is no need to trigger a migration manually.

You _can_ trigger migrations manually if you really want (and I can show you how to do it), but I hope regular drift migrations should work for you since they're considerably easier to use. So if they haven't been working for you in the server case, maybe that's something I need to know and look into :)

**Reply by wstrange** [reply]
I did read the docs (honest!) but changing the schemaVersion has no impact. The upgrade would never get triggered.   

Any suggestions on debugging this on Mac OS? I'm using the NativeDatabase instance.  Does the SQL browser offer a way to inspect the header?

**Comment by wstrange** [comment]
I found the SQL editor can edit/show pragmas, including the user version


If I shut down my app, bump the `schemaVersion => 3`, and start up again, the schema does not increment in the pragma (and the upgrade does not run). 

Am I doing something daft?

Here is a more complete snippet of how I open the database:

```dart
class Database extends _$Database {
  // we tell the database where to store the data with this constructor
  Database() : super(_openConnection());
  
  @override
  int get schemaVersion => 4;

  MigrationStrategy get migration {
    return MigrationStrategy(onCreate: (Migrator m) async {
      await m.createAll();
    },
    onUpgrade: (Migrator m, int from, int to) async {
      print('Migrate from $from to $to');
        await m.createTable(notificationTokenTbl);
    },
    beforeOpen: (details) async {
      print('before open');
    });
  }
}

LazyDatabase _openConnection() {
  // the LazyDatabase util lets us find the right location for the file async.
  return LazyDatabase(() async {
   
    final db = '/Users/warren/tmp/db.sqlite';
    final file = File(db);
    // file.deleteSync();
    return NativeDatabase.createInBackground(file, logStatements: true);
  });
}



```

**Reply by simolus3** [reply]
Your snippet looks correct to me. Could you print the details in the `beforeOpen` callback:

```dart
  beforeOpen: (details) async {
    print('before open: ${details.versionBefore} -> ${details.versionNow}');
  },
```

Does this show the same version? Is `versionNow` your current `schemaVersion`?

-------

If you want to take a look at what's going in drift, the relevant snippet is this one

https://github.com/simolus3/drift/blob/e433cff932f82c97ecc365a4ec5e456b0b68f236/drift/lib/src/runtime/executor/helpers/engines.dart#L440-L458

The `DynamicVersionDelegate` is relevant for native databases, and `versionDelegate.schemaVersion` will load the user version internally. The call in line 458 will tell the database the old schema version, it will react to it by running a migration if needed.

**Reply by wstrange** [reply, marked as answer]
The print statement was not reached, but I think this is something to do with how I initialize the Database class.

I had all my queries in a dbService class, and initialized the database like this:

```dart
// top level - not in a class
final db = Database();
```

I gather the Dart initializer for top level finals somehow runs before the file can be opened (it is lazy opened). If I move the initialization into my main(), it is now triggering the upgrade.

Thanks for the clues to get me there!

**Reply by simolus3** [reply]
Hm, that's still weird since Dart fields are semantically initialized when first accessed (otherwise they wouldn't be able to get tree-shaken when unused).

Drift only opens the database and runs the migrations when the first query is executed. So perhaps the problem was that the database was never fully opened? Glad you were able to resolve the issue though!

**Reply by wstrange** [reply]
> So perhaps the problem was that the database was never fully opened? 

Yea, the `final var` thing may not be the issue. I had to add a sample query in main() to trigger the migration.  It's working now - so all good.
